File: Introductory/gridPath/gridPath.cpp
#include <iostream>
#include <string>
using namespace std;

string path;
int n;
bool visited[7][7];
long long ways = 0;

bool blocked(int i, int j) {
    return i < 0 || j < 0 || i >= 7 || j >= 7 || visited[i][j];
}

// corridor pruning: returns true if continuing from (i, j) would split the
// remaining free region into two parts that can't both be filled.
bool splits_grid(int i, int j) {
    bool up    = blocked(i - 1, j);
    bool down  = blocked(i + 1, j);
    bool left  = blocked(i, j - 1);
    bool right = blocked(i, j + 1);

    // straight pinch: blocked on two opposite sides, open on the other two
    if ((up && down && !left && !right) ||
        (left && right && !up && !down))
        return true;

    // diagonal pinch: a blocked diagonal whose two connecting sides are both
    // open splits the remaining free region into two halves
    if (blocked(i - 1, j - 1) && !up && !left)   return true;
    if (blocked(i - 1, j + 1) && !up && !right)  return true;
    if (blocked(i + 1, j - 1) && !down && !left) return true;
    if (blocked(i + 1, j + 1) && !down && !right) return true;

    return false;
}

const int DI[4] = {-1, 0, 1, 0};   // U, L, D, R
const int DJ[4] = {0, -1, 0, 1};

void recurse(int move, int i, int j) {
    // reached the lower-left corner
    if (i == 6 && j == 0) {
        if (move == n) ways++;   // valid only if all 49 squares visited
        return;
    }

    if (move >= n) return;
    if (blocked(i, j)) return;
    if (splits_grid(i, j)) return;

    visited[i][j] = true;

    char c = path[move];
    for (int d = 0; d < 4; d++) {
        // which directions does this character allow?
        bool allow = (c == '?')
                  || (c == 'U' && d == 0)
                  || (c == 'L' && d == 1)
                  || (c == 'D' && d == 2)
                  || (c == 'R' && d == 3);
        if (!allow) continue;

        int ni = i + DI[d], nj = j + DJ[d];
        if (!blocked(ni, nj))
            recurse(move + 1, ni, nj);
    }

    visited[i][j] = false;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> path;
    n = (int)path.size();

    recurse(0, 0, 0);
    cout << ways << "\n";
    return 0;
}

File: Introductory/gridPath/gridPath.py
ways = 0

moves = {
    '?': [(-1, 0), (0, -1), (1, 0), (0, 1)],
    'U': [(-1, 0)],
    'D': [(1, 0)],
    'L': [(0, -1)],
    'R': [(0, 1)]
}

def blocked(i, j, visitedSet):
    return i < 0 or j < 0 or i >= 7 or j >= 7 or (i, j) in visitedSet

def splits_grid(i, j, visitedSet):
    up = blocked(i - 1, j, visitedSet)
    down = blocked(i + 1, j, visitedSet)
    left = blocked(i, j - 1, visitedSet)
    right = blocked(i, j + 1, visitedSet)

    # straight pinch: blocked on two opposite sides, open on the other two
    if (up and down and not left and not right) or \
       (left and right and not up and not down):
        return True

    # diagonal pinch: a blocked diagonal whose two connecting sides are both
    # open splits the remaining free region into two halves
    if blocked(i - 1, j - 1, visitedSet) and not up and not left: return True
    if blocked(i - 1, j + 1, visitedSet) and not up and not right: return True
    if blocked(i + 1, j - 1, visitedSet) and not down and not left: return True
    if blocked(i + 1, j + 1, visitedSet) and not down and not right: return True

    return False

def recurse(inpList, move, i, j, visitedSet):
    global ways

    if i == 6 and j == 0:
        if move == len(inpList):
            ways += 1
        return

    if move >= len(inpList):
        return

    if blocked(i, j, visitedSet):
        return

    if splits_grid(i, j, visitedSet):
        return

    visitedSet.add((i, j))

    for di, dj in moves[inpList[move]]:
        ni, nj = i + di, j + dj

        if not blocked(ni, nj, visitedSet):
            recurse(inpList, move + 1, ni, nj, visitedSet)

    visitedSet.remove((i, j))

if __name__ == "__main__":
    inpList = list(input())
    recurse(inpList, 0, 0, 0, set())
    print(ways)
